File: java/Patterns/src/com/bilak/structural/facade/RentalFacade.java
package com.bilak.structural.facade;

import com.bilak.domain.EquipmentCategory;
import com.bilak.structural.facade.domain.Client;
import com.bilak.structural.facade.domain.Equipment;
import com.bilak.structural.facade.subcomponents.EquipmentService;
import com.bilak.structural.facade.subcomponents.PaymentService;
import com.bilak.structural.facade.subcomponents.RentalService;

import java.time.LocalDateTime;

/**
 * @author Yuliana
 * @version 1.0.0
 * @group 343(1)
 * @department PZKS
 * @project java
 * @class RentalFacade
 * @since 10/14/2022 - 11.21
 **/
public class RentalFacade {
    public boolean makeRental(Client client, String equipmentType, EquipmentCategory category) {
        boolean rentalFulfilled = false;
        Equipment eqForRent = EquipmentService.findEquipment(client.getHeight(), equipmentType, category);
        if (eqForRent != null) {
            System.out.println("Equipment " + eqForRent.getName() + " " + eqForRent.getCategory() + " is available.");
            if (PaymentService.makePayment(eqForRent.getBasePrice(), client)) {
                System.out.println("Payment confirmed.");
                RentalService.add("Rental at " + LocalDateTime.now() + " for client " + client + " with equipment " + eqForRent);
                rentalFulfilled = true;
            }
        }
        return rentalFulfilled;
    }
}

File: javascript/structural/facade/rental-facade.js
import { EquipmentService } from "./equipment-service.js";
import { PaymentService } from "./payment-service.js";
import { RentalService } from "./rental-service.js";

export class RentalFacade {
    constructor() {
        this.equipmentService = new EquipmentService();
        this.paymentService = new PaymentService();
        this.rentalService = new RentalService();
    }

    makeRental(client, equipmentType, category) {
        let rentalFulfilled = false;
        const eqForRent = this.equipmentService.findEquipment(client.height, equipmentType, category);
        if (eqForRent) {
            console.log(`Equipment ${eqForRent.name} ${eqForRent.category} is available.`);
            if (this.paymentService.makePayment(eqForRent.basePrice, client)) {
                console.log("Payment confirmed.");
                this.rentalService.add(`Rental at ${new Date().toISOString()} for client ${client} with equipment ${eqForRent}`);
                rentalFulfilled = true;
            }
        }
        return rentalFulfilled;
    }
}